File: web-app/src/services/NewsService.ts
import { LruCache } from '@/utils/LruCache';
import { ApiQuotaLedger } from '@/utils/ApiQuotaLedger';
import { logApiCall } from '@/utils/logMetrics';
import { NetClient, HALF_DAY_MS } from '../../../packages/core/net';
/* global DOMParser */

export interface NewsArticle {
  title: string;
  url: string;
  source: string;
  published: string;
}

interface NewsApiEntry {
  title: string;
  link: string;
  source_id: string;
  pubDate: string;
}

/**
 * Service for fetching recent news articles for a stock symbol.
 */
export class NewsService {
  private cache = new LruCache<string, NewsArticle[]>(32);
  private ledger = new ApiQuotaLedger(200); // 200 req/day
  private client = new NetClient(this.ledger);
  private apiKey: string;
  constructor(apiKey: string) {
    if (typeof apiKey !== 'string' || apiKey.trim() === '') {
      throw new Error('Newsdata API key is required');
    }
    this.apiKey = apiKey;
  }

  private parseRss(xml: string): NewsArticle[] {
    const doc = new DOMParser().parseFromString(xml, 'text/xml');
    const items = Array.from(doc.querySelectorAll('item')).slice(0, 3);
    return items.map(item => ({
      title: item.querySelector('title')?.textContent ?? '',
      url: item.querySelector('link')?.textContent ?? '',
      source: item.querySelector('source')?.textContent ?? 'rss',
      published: item.querySelector('pubDate')?.textContent ?? '',
    }));
  }

  /**
   * Retrieve up to three recent articles about the given symbol.
   *
   * API usage is accounted for using the quota ledger and results
   * are cached for twelve hours.
   *
   * @param symbol - Stock ticker symbol.
   * @returns A list of news articles or `null` on failure.
   */
  async getNews(symbol: string): Promise<NewsArticle[] | null> {
    const url = `https://newsdata.io/api/1/news?apikey=${this.apiKey}&q=${symbol}&language=en`;
    const start = performance.now();
    if (!this.ledger.isSafe()) return null;
    let articles = await this.client.get<NewsArticle[]>(
      url,
      this.cache,
      json => {
        const results: NewsApiEntry[] = json.results ?? [];
        return results.slice(0, 3).map(({ title, link, source_id, pubDate }) => ({
          title,
          url: link,
          source: source_id,
          published: pubDate,
        }));
      },
      HALF_DAY_MS
    );
    logApiCall('NewsService.getNews', start);
    if (articles) return articles;

    const rssUrl = 'https://rss.theguardian.com/business/markets/index.xml';
    const cached = this.cache.get(rssUrl);
    if (cached !== undefined) return cached;
    try {
      const resp = await fetch(rssUrl);
      if (!resp.ok) return null;
      const xml = await resp.text();
      articles = this.parseRss(xml);
      this.cache.put(rssUrl, articles, HALF_DAY_MS);
      return articles;
    } catch {
      return null;
    }
  }
}

File: mobile-app/packages/services/lib/src/news_service.dart
import 'lru_cache.dart';
import 'api_quota_ledger.dart';
import 'fetch_json.dart';
import 'package:http/http.dart' as http;
import 'package:xml/xml.dart';

/// S-03 – NewsService
class NewsService {
  final LruCache<String, List<Map<String, dynamic>>> _cache = LruCache(32);
  final LruCache<String, List<Map<String, dynamic>>> _rssCache = LruCache(4);
  final ApiQuotaLedger _ledger;
  final http.Client _client;
  late final NetClient _net;
  final String _apiKey;
  static const _rssUrl =
      'https://rss.theguardian.com/business/markets/index.xml';

  NewsService(this._apiKey, {ApiQuotaLedger? ledger, http.Client? client})
      : _ledger = ledger ?? ApiQuotaLedger(200),
        _client = client ?? http.Client() {
    if (_apiKey.trim().isEmpty) {
      throw ArgumentError('Newsdata API key is required');
    }
    _net = NetClient(_ledger, _client);
  }

  Future<List<Map<String, dynamic>>?> getDigest(String topic) async {
    if (!_ledger.isSafe()) {
      return null;
    }
    final url =
        'https://newsdata.io/api/1/news?apikey=$_apiKey&q=$topic&language=en';
    final articles = await _net.get<List<Map<String, dynamic>>>(
      url,
      _cache,
      (json) {
        final results = json['results'] as List? ?? [];
        return results.take(3).map((n) {
          return {
            'title': n['title'],
            'url': n['link'],
            'source': n['source_id'],
            'published': n['pubDate'],
          };
        }).toList();
      },
      ttl: const Duration(hours: 12),
    );
    return articles ?? await _fetchRss();
  }

  Future<List<Map<String, dynamic>>?> _fetchRss() async {
    final cached = _rssCache.get(_rssUrl);
    if (cached != null) return cached;
    try {
      final resp = await _client.get(Uri.parse(_rssUrl));
      if (resp.statusCode != 200) return null;
      final doc = XmlDocument.parse(resp.body);
      final items = doc.findAllElements('item').take(3).map((n) {
        return {
          'title': n.getElement('title')?.text ?? '',
          'url': n.getElement('link')?.text ?? '',
          'source': 'rss',
          'published': n.getElement('pubDate')?.text ?? '',
        };
      }).toList();
      _rssCache.put(_rssUrl, items, const Duration(hours: 12));
      return items;
    } catch (_) {
      return null;
    }
  }
}
